buffer/src/main/java/io/netty/buffer/SimpleLeakAwareByteBuf.java

```diff
@@ -16,64 +16,93 @@
 
 package io.netty.buffer;
 
-import io.netty.util.ResourceLeak;
+import io.netty.util.ResourceLeakDetector;
+import io.netty.util.ResourceLeakTracker;
+import io.netty.util.internal.ObjectUtil;
 
 import java.nio.ByteOrder;
 
-final class SimpleLeakAwareByteBuf extends WrappedByteBuf {
+class SimpleLeakAwareByteBuf extends WrappedByteBuf {
 
-    private final ResourceLeak leak;
+    /**
+     * This object's is associated with the {@link ResourceLeakTracker}. When {@link ResourceLeakTracker#close(Object)}
+     * is called this object will be used as the argument. It is also assumed that this object is used when
+     * {@link ResourceLeakDetector#track(Object)} is called to create {@link #leak}.
+     */
+    private final ByteBuf trackedByteBuf;
+    final ResourceLeakTracker<ByteBuf> leak;
 
-    SimpleLeakAwareByteBuf(ByteBuf buf, ResourceLeak leak) {
-        super(buf);
-        this.leak = leak;
+    SimpleLeakAwareByteBuf(ByteBuf wrapped, ByteBuf trackedByteBuf, ResourceLeakTracker<ByteBuf> leak) {
+        super(wrapped);
+        this.trackedByteBuf = ObjectUtil.checkNotNull(trackedByteBuf, "trackedByteBuf");
+        this.leak = ObjectUtil.checkNotNull(leak, "leak");
+    }
+
+    SimpleLeakAwareByteBuf(ByteBuf wrapped, ResourceLeakTracker<ByteBuf> leak) {
+        this(wrapped, wrapped, leak);
+    }
+
+    @Override
+    public ByteBuf slice() {
+        return newSharedLeakAwareByteBuf(super.slice());
+    }
+
+    @Override
+    public ByteBuf slice(int index, int length) {
+        return newSharedLeakAwareByteBuf(super.slice(index, length));
+    }
+
+    @Override
+    public ByteBuf duplicate() {
+        return newSharedLeakAwareByteBuf(super.duplicate());
+    }
+
+    @Override
+    public ByteBuf readSlice(int length) {
+        return newSharedLeakAwareByteBuf(super.readSlice(length));
     }
 
     @Override
     public boolean release() {
-        boolean deallocated =  super.release();
-        if (deallocated) {
-            leak.close();
+        if (super.release()) {
+            closeLeak();
+            return true;
         }
-        return deallocated;
+        return false;
     }
 
     @Override
     public boolean release(int decrement) {
-        boolean deallocated = super.release(decrement);
-        if (deallocated) {
-            leak.close();
+        if (super.release(decrement)) {
+            closeLeak();
+            return true;
         }
-        return deallocated;
+        return false;
+    }
+
+    private void closeLeak() {
+        // Close the ResourceLeakTracker with the tracked ByteBuf as argument. This must be the same that was used when
+        // calling DefaultResourceLeak.track(...).
+        boolean closed = leak.close(trackedByteBuf);
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Remove this useless assignment to local variable "closed". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1854)